enhancements/dns/coredns-egressfirewall-integration.md

```diff
@@ -0,0 +1,363 @@
+
+
+### Risks and Mitigations
+
+
+### Drawbacks
+
+* Whenever there's a change in the IPs or the next lookup time for a DNS name, the additional step of updating the
+related ``DNSName`` CRs will be executed. This might add some delay to the DNS lookup process. However, this will only
+happen whenever there's a change in the DNS information.
+
+
+## Design Details
+
+### Open Questions [optional]
+
+### Test Plan
+
+* This enhancement will be tested through e2e tests by adding EgressFirewall rules containing regular DNS names
+and wildcard DNS names.
+* Testing the feature where IPs are dynamically changed may be a little bit tricky as this will probably include
+creation of DNS records and then changing the IP adresses of the DNS records through the e2e tests.
+
+### Graduation Criteria
+
+This is a user facing change and will directly go to GA. This feature requires an update to Openshift Docs.
+
+#### Dev Preview -> Tech Preview
+
+N.A. This feature will go directly to GA.
+
+#### Tech Preview -> GA
+
+N.A. This feature will go directly to GA.
+
+#### Removing a deprecated feature
+
+
+### Upgrade / Downgrade Strategy
+
+Upgrade expectations:
+* On upgrade, the OVN-K master will create the corresponding ``DNSName`` CRs for each DNS name in the
+existing EgressFirewall resources. The ``egressfirewall`` plugin will also start updating the ``.status``
+fields of the ``DNSName`` CRs. The scenarios arising out of the order of the update of the various components
+are dicussed in [Version Skew Strategy](#version-skew-strategy)
+
+Downgrade expectations:
+* On downgrade, the ``DNSName`` CRs may still remain. However, these CRs would not have any impact on how
+EgressFirewall ACL rules are implemented in the downgraded cluster. Deleting the CR Definition of ``DNSName``
+from the cluster would remove all the ``DNSName`` CRs.
+
+### Version Skew Strategy
+
+The following 2 scenarios may occur during the upgrade process:
+* Scenario 1: The Cluster DNS operator and the CoreDNS pods are upgraded first and then the OVN-K master pods.
+
+  In this scenario, the ``egressfirewall`` CoreDNS plugin will start inspecting each DNS lookup before the ``DNSName``
+  CRs are created by the OVN-K master. The plugin will just respond with the response received from other plugins for
+  the DNS lookups. As OVN-K master will be continuing the DNS lookups for DNS names with expired TTLs, CoreDNS will
+  also be responding with the corresponding IPs and the TTLs.
+
+* Scenario 2: The OVN-K master pods are upgraded first and then the Cluster DNS operator and the CoreDNS pods.
+
+  In this scenario, the OVN-K master will create ``DNSName`` CRs for each unique DNS name used in EgressFirewall rules.
+  However, as the Cluster DNS operator and the CoreDNS pods are still not upgraded, CoreDNS pods will not run the
+  ``egressfirewall`` plugin. The OVN-K master will still receive the response for the DNS lookup queries it will send
+  to the CoreDNS.
```

Comment:
...which means that egressfirewall DNS rules will not work correctly until after CoreDNS is updated

Reply:
The rules will function as it did before the upgrade as the CoreDNS will still respond to the DNS lookups done by the OVN-K master. The new features like supporting wildcard DNS names and supporting dynamically changing IP addresses will not work. But the previous support of egressfirewall DNS rules will still be there.

Reply:
The previous support included its own solution for dynamically changing IP addresses. Assuming that we remove the old code from ovn-kubernetes, then during the skew period, neither new ovn-k nor old coredns would be handling dynamic IPs.